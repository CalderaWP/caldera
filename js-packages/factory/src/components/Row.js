import React, {Fragment} from 'react';
import {Fields} from './Fields';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {Flex, Box} from '@rebass/grid'


/**
 * Render a row of fields with a shared onChange/onBlur event for all values
 * @param columns
 * @param onChange
 * @param onBlur
 * @param className
 * @param rowId
 * @return {*}
 * @constructor
 */
export const Row = ({
						columns, onChange, onBlur, className, rowId
					}) => (
	<Flex
		className={classNames('caldera-row', className)}
		id={rowId}
	>
		{columns.map(column => {
			let {
				width,
				padding,
				columnId,
				fields,
			} = column;

			padding = padding ? padding : 8;

			return (
				<Box
					key={columnId}
					id={columnId}
					width={width}
					px={padding}
					py={padding}
				>
					<Fields
						fields={fields}
						onChange={onChange}
						onBlur={onBlur}
					/>
				</Box>

			)
		})}
	</Flex>

);

/**
 * Row prop types
 *
 * @type {{columns: shim, rowId: *, onChange: *, onBlur: shim, className: shim}}
 */
export const rowPropTypes = {
	columns: PropTypes.arrayOf(PropTypes.shape({
		fields: PropTypes.array.isRequired,
		width: PropTypes.string.isRequired,
		columnId: PropTypes.oneOfType([
			PropTypes.string,
			PropTypes.number,
		])
		//input propTypes?
	})),
	rowId: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.number,
	]),
	onChange: PropTypes.func.isRequired,
	onBlur: PropTypes.func,
	className: PropTypes.string
};
Row.propTypes = rowPropTypes;
