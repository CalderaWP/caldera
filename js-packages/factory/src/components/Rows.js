import React, {Fragment} from 'react';
import {Row, rowPropTypes} from './Row';
import PropTypes from 'prop-types';

export const Rows = ({
						 rows, onChange, onBlur, className
					 }) => (
	<Fragment>
		{rows.forEach(row => <Row
			key={row.rowId}
			{...row}
			onChange={onChange}
			onBlur={onBlur}/>)
		}
	</Fragment>
);

Rows.propTypes = {
	rows: PropTypes.arrayOf(PropTypes.shape(rowPropTypes)),
	onChange: PropTypes.func.isRequired,
	onBlur: PropTypes.func,
	className: PropTypes.string
};
