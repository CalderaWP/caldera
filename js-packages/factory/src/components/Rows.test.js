import React from 'react';
import renderer from 'react-test-renderer';
import {mount} from 'enzyme';
import {
	checkboxFieldset,
	selectField,
	checkboxField,
	numberField,
	textField,
	emailField
} from '../fields.fixtures';

import {Rows} from './Rows';
import {Row} from "./Row";

describe('Row component', () => {
	let onChange;
	let onBlur;
	beforeEach(() => {
		onChange = jest.fn();
		onBlur = jest.fn()
	});
	it('Creates with two rows', () => {

		const row1 = {
			rowId: 'r1',
			onChange,
			onBlur,
			columns: [
				{
					fields: [
						textField,
						checkboxField,
					],
					width: '1/2'
				},
				{
					fields: [
						checkboxFieldset,
					],
					width: '1/2'
				}
			]
		};

		const row2 = {
			rowId: 'r2',
			onChange,
			onBlur,
			columns: [
				{
					fields: [
						selectField,
					],
					width: '1/3'
				},
				{
					fields: [
						emailField,
					],
					width: '1/3'
				},
				{
					fields: [
						checkboxField,
					],
					width: '1/3'
				}
			]
		};



		const component = renderer.create(
			<Rows
				rows={[row1,row2]} onChange={onChange} onBlur={onBlur}
			/>
		);
		expect(component.toJSON()).toMatchSnapshot();
	});


});
