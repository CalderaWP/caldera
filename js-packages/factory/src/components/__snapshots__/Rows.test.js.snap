// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Row component Creates with two rows 1`] = `null`;
